src/ripple/nodestore/impl/DatabaseShardImp.h

```diff
@@ -0,0 +1,164 @@
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_NODESTORE_DATABASESHARDIMP_H_INCLUDED
+#define RIPPLE_NODESTORE_DATABASESHARDIMP_H_INCLUDED
+
+#include <ripple/nodestore/DatabaseShard.h>
+#include <ripple/nodestore/impl/Shard.h>
+
+namespace ripple {
+namespace NodeStore {
+
+class DatabaseShardImp : public DatabaseShard
+{
+public:
+    DatabaseShardImp() = delete;
+    DatabaseShardImp(DatabaseShardImp const&) = delete;
+    DatabaseShardImp& operator=(DatabaseShardImp const&) = delete;
+
+    DatabaseShardImp(Application& app, std::string const& name,
+        Stoppable& parent, Scheduler& scheduler, int readThreads,
+            Section const& config, beast::Journal j);
+
+    ~DatabaseShardImp() override;
+
+    bool
+    init() override;
+
+    boost::optional<std::uint32_t>
+    prepare(std::uint32_t validLedgerSeq) override;
+
+    std::shared_ptr<Ledger>
+    fetchLedger(uint256 const& hash, std::uint32_t seq) override;
+
+    void
+    setStored(std::shared_ptr<Ledger const> const& ledger) override;
+
+    bool
+    hasLedger(std::uint32_t seq) override;
+
+    std::string
+    getCompleteShards() override;
+
+    void
+    validate() override;
+
+    std::string
+    getName() const override
+    {
+        return "shardstore";
+    }
+
+    void
+    for_each(std::function <void(std::shared_ptr<NodeObject>)> f) override
+    {
+        Throw<std::runtime_error>("Shard store import not supported");
+    }
+
+    void
+    import(Database& source) override
+    {
+        Throw<std::runtime_error>("Shard store import not supported");
+    }
+
+    std::int32_t
+    getWriteLoad() const override;
+
+    void
+    store(NodeObjectType type, Blob&& data,
+        uint256 const& hash, std::uint32_t seq) override;
+
+    std::shared_ptr<NodeObject>
+    fetch(uint256 const& hash, std::uint32_t seq) override;
+
+    bool
+    asyncFetch(uint256 const& hash, std::uint32_t seq,
+        std::shared_ptr<NodeObject>& object) override;
+
+    bool
+    copyLedger(std::shared_ptr<Ledger const> const& ledger) override;
+
+    int
+    getDesiredAsyncReadCount(std::uint32_t seq) override;
+
+    float
+    getCacheHitRate() override;
+
+    void
+    tune(int size, int age) override;
+
+    void
+    sweep() override;
+
+private:
+    Application& app_;
+    mutable std::mutex m_;
+    std::map<std::uint32_t, std::unique_ptr<Shard>> complete_;
+    std::unique_ptr<Shard> incomplete_;
+    Section const config_;
+    boost::filesystem::path dir_;
+
+    // If new shards can be stored
+    bool canAdd_ {true};
+
+    // Complete shard indexes
+    std::string status_;
+
+    // If backend type uses permanent storage
+    bool backed_;
+
+    // Maximum disk space the DB can use (in bytes)
```

Comment:
If the backend does not use permanent storage, the maximum sizes below are not enforced, so it might use unlimited memory?

Reply:
Yes, and the `Backend` interface does not provide info about memory consumption, luckily we only use it for unit tests.